src/service/rbac-service.ts

```diff
@@ -17,42 +17,246 @@
+      };
+    });
+  }
+
+  /**
+   * Convert a human-readable permission definition to a list of
+   * database permission rules
+   * @param def
+   */
+  public static definitionToRules(def: PermissionDefinition): PermissionRule[] {
+    const permissions: PermissionRule[] = [];
+    Object.keys(def).forEach(entity => {
+      Object.keys(def[entity]).forEach((action) => {
+        Object.keys(def[entity][action]).forEach((relation) => {
+          const attributes = Array.from(def[entity][action][relation]);
+          permissions.push({ entity, action, relation, attributes });
+        });
+      });
+    });
+    return permissions;
+  }
+
+  /**
+   * Convert a list of database permission rules to a human-readable
+   * permissions object
+   * @param rules
+   */
+  public static rulesToDefinition(rules: PermissionRule[]): PermissionDefinition {
+    return rules.reduce((permDef: PermissionDefinition, permission1, i1, rolePermissions) => {
+      if (permDef[permission1.entity]) return permDef;
+
+      permDef[permission1.entity] = rolePermissions
+        .filter((p) => p.entity === permission1.entity)
+        .reduce((entDef: EntityDefinition, permission2, i2, entityPermissions) => {
+          if (entDef[permission2.entity]) return entDef;
+
+          entDef[permission2.action] = entityPermissions
+            .filter((p) => p.action === permission2.action)
+            .reduce((actDef: ActionDefinition, permission3) => {
+              if (actDef[permission3.relation]) return;
+
+              actDef[permission3.relation] = new Set(permission3.attributes);
+              return actDef;
+
+            }, {});
+          return entDef;
+
+        }, {});
+      return permDef;
+    }, {});
+  }
+
+  /**
+   * Get a tuple with a list of all roles and the total number of
+   * roles matching the parameters
+   * @param params
+   * @param take
+   * @param skip
+   */
+  public static async getRoles(params: RoleFilterParameters = {}, { take, skip }: PaginationParameters = {}): Promise<[Role[], number]> {
+    const options = this.getOptions(params);
+    return Role.findAndCount({ ...options, take, skip });
+  }
+
+  /**
+   * Create an new role with the given parameters
+   * @param params
+   */
+  public static async createRole(params: UpdateRoleParams) {
+    return Role.save({ ...params, permissions: [] });
+  }
+
+  /**
+   * Update an existing role with the given parameters
+   * @param roleId
+   * @param params
+   */
+  public static async updateRole(roleId: number, params: UpdateRoleParams): Promise<Role> {
+    const role = await Role.findOne({ where: { id: roleId } });
+    if (role == null) throw new Error('Role not found.');
+    if (role.systemDefault) throw new Error('Cannot update system default role.');
+
+    role.name = params.name;
+    await role.save();
+
+    const [[r]] = await this.getRoles({ roleId, returnPermissions: true });
```

Comment:
still kinky

Reply:
See the comment above about pagination for this method.